### PR TITLE
WIP/TODO stream/pdu: initial stream PDU support

### PR DESCRIPTION
**Pull request by victorjulien**
The idea of stream PDUs is that the applayer parsers can tag PDUs
in the stream while parsing. These PDUs can then be inspected from the
rule language. This will allow rules that are more precise and less
costly.

Consider a rule like:

alert tcp ... flow:to_server; content:"|ff|SMB"; content:"some smb 1 issue";

this will scan all toserver TCP traffic, where it will only be limited by a port,
depending on how rules are grouped.

With this work we'll be able to do:

alert smb ... flow:to_server; smb1.data; content:"some smb 1 issue";

This rule will only inspect the data portion of SMB1 records. It will not affect
any other protocol, and it won't need special patterns to "search" for the
SMB1 record in the raw stream.

An EVE PDU logger was added. This is mostly to assist development and QA. It
produces too much data for practical use.

PDUs have / should get events to inspect/log parser errors that don't fit the
TX model.

PDUs are stored in `TcpStream::pdus`, with a fixed size static portion and
dynamic support for a larger number. This is done for effeciency but may be
a case of too early optimizations.

When the Stream Buffer slides, records are updated as they use offsets relative
to the stream. A negative offset is used for StreamPDUs that started before the
current window.

TODO

Current PoC implementation tracks StreamPDUs in the `TcpStream` structure, but
as its fairly large it may need to go into the AppLayerParserState or per
app-parser state.

PDUs with unknown size.

PDUs that just run till the end of the stream.

PDUs in alerts

PDU support in all the protocols

Avoiding rescanning of the same PDU.

Clean up debug cruft

Mapping between per alproto record types and printable names.

Implement per PDU events.

Condider the detection side: should we create a keyword per buffer or
one unified one that can do something like `pdu:smb1.data;` where the
types can be registered from the parser? Probably yes.

Record creation funcs currently return a ptr to the StreamPDU, but these
pointers should not be kept as after a slide they are invalid. Need a
different idenfifier.

Many more things.

### CODE REVIEW

File: rust/src/smb/smb.rs [hunk 1]
```diff
@@ -53,6 +53,23 @@ use crate::smb::events::*;
 use crate::smb::files::*;
 use crate::smb::smb2_ioctl::*;
 
+#[repr(C)]
+pub enum SMBPduType {
```
Comment (victorjulien) [inline rust/src/smb/smb.rs:57]: on question about the implementation is how much granularity we need to add wrt to the record types. Does it make sense to split by SMB version here?

**Review by jasonish — COMMENTED** [review]
> Condider the detection side: should we create a keyword per buffer or
> one unified one that can do something like `pdu:smb1.data;` where the
> types can be registered from the parser? Probably yes.

Yes, this would be nice.

File: rust/src/smb/smb.rs [hunk 2]
```diff
@@ -1511,6 +1544,14 @@ impl SMBState {
 
         match parse_nbss_record_partial(input) {
             Ok((output, ref nbss_part_hdr)) => {
+
+                let _nbss_pdu = applayer_new_record_tc(flow, base, input, nbss_part_hdr.length as i32 + 4, SMBPduType::NBSS as u8);
```
Comment (jasonish) [inline rust/src/smb/smb.rs:1548]: Why would the parser need a "handle" for the record?

File: rust/src/smb/smb.rs [hunk 3]
```diff
@@ -53,6 +53,23 @@ use crate::smb::events::*;
...
+    SMB1 = 3,
+    SMB1Hdr = 4,
+    SMB1Data = 5,
+    SMB2 = 6,
+    SMB2Hdr = 7,
+    SMB2Data = 8,
+    SMB3 = 9,
+    SMB3Hdr = 10,
+    SMB3Data = 11,
+}
```
[the remark this answers is not in the document]
Comment (jasonish) [reply rust/src/smb/smb.rs:70]: Maybe something like:
```rust
enum SmbRecordType {
    NBSSPdu,
    NBSSHdr,
    NBSSData,
}
```

File: rust/src/smb/smb.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/smb/smb.rs:1548]: To update it later to:
- set the tx_id
- set an event if the parser encounters an exception in parsing the record

File: rust/src/smb/smb.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/smb/smb.rs:70]: yeah good points. In a upcoming draft I did something like
`record:tls.pdu;`
`record:tls.hdr;`
We can have a convention that `pdu` means a whole record.

Thinking about the rule option name. Maybe `app-record:tls.pdu;` would make sense.

[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/smb/smb.rs:70]: Or if we choose "frame" something like `app-frame:tls.pdu;`.

[the remark this answers is not in the document]
Comment (jasonish) [reply rust/src/smb/smb.rs:70]: Is the `app` in `app-frame` a bit redundant?  The `tls` implies it comes from the TLS.  What other kind of frames might there be that conflict with `frame:<protoname>.<framename>`?

**Comment by suricata-qa** [comment]
ERROR: QA failed on tlpr1_autofp_suri.

Pipeline 4344

**Comment by victorjulien** [comment]
Replaced by #6664

**Pull request closed without merging**